src/librustc_metadata/cstore_impl.rs

```diff
@@ -239,7 +241,30 @@ provide! { <'tcx> tcx, def_id, other, cdata,
 
     used_crate_source => { Lrc::new(cdata.source.clone()) }
 
-    exported_symbols => { Arc::new(cdata.exported_symbols(tcx)) }
+    exported_symbols => {
+        let mut syms = cdata.exported_symbols(tcx);
+
+        // When linked into a dylib crates don't export their generic symbols,
+        // so if that's happening then we can't load upstream monomorphizations
+        // from this crate. As a result, if we're creating a dylib or this crate
+        // is being included from a different dynamic library, then we filter
+        // out all `Generic` symbols here.
+        let formats = tcx.dependency_formats(LOCAL_CRATE);
+        let remove_generics = formats.iter().any(|(ty, list)| {
+            *ty == CrateType::Dylib ||
```

Comment:
This might be a bit more restrictive than necessary. If the current crate is a dylib, it can still link against upstream generics. It just shouldn't re-export them.

Reply:
True yeah, and I've added a commit now which is relatively low-impact which should solve this